File: python/111_minimum-depth-of-binary-tree.py
from collections import deque

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

# DFS with Recursion
class Solution:
    def minDepth(self, root: TreeNode) -> int:
        return self.recursion(root, 1, float('inf'))

    def recursion(self, root, d, ans):
        if not root:
            return 0
        if not root.left and not root.right:
            return min(d, ans)

        if root.left:
            ans = self.recursion(root.left, d+1, ans)
        if root.right:
            ans = self.recursion(root.right, d+1, ans)

        return ans


#---------------------------------------------------------------------------------------#


# DFS with Stack
class Solution:
    def minDepth(self, root: TreeNode) -> int:
        if not root:
            return 0

        ans = float('inf')
        stack = deque()
        stack.append((root, 1))

        count = 0
        while len(stack) > 0:
            node, depth = stack.pop()

            if not node.left and not node.right:
                ans = min(ans, depth)
            if node.left:
                stack.append((node.left, depth + 1))
            if node.right:
                stack.append((node.right, depth + 1))

        return ans


#---------------------------------------------------------------------------------------#


# BFS with Queue
class Solution:
    def minDepth(self, root: TreeNode) -> int:
        if not root:
            return 0

        ans = float('inf')
        queue = deque()
        queue.append((root, 1))

        count = 0
        while len(queue) > 0:
            node, depth = queue.popleft()

            if not node.left and not node.right:
                ans = min(ans, depth)
            if node.left:
                queue.append((node.left, depth + 1))
            if node.right:
                queue.append((node.right, depth + 1))

        return ans

File: go/111_minimum-depth-of-binary-tree.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

type queue struct {
	node *TreeNode
	l    int
}

func minDepth(root *TreeNode) int {
	if root == nil {
		return 0
	}

	var q []queue
	q = append(q, queue{root, 1})

	for len(q) > 0 {
		node, l := q[0].node, q[0].l
		q = q[1:]

		if node.Left == nil && node.Right == nil {
			return l
		}

		if node.Left != nil {
			q = append(q, queue{node.Left, l + 1})
		}
		if node.Right != nil {
			q = append(q, queue{node.Right, l + 1})
		}
	}
	return -1
}
